x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java

```diff
@@ -218,18 +239,88 @@ private void match(int stage, SequenceKey key, Ordinal ordinal, HitReference hit
                 }
             }
 
-            completed.add(sequence);
+            tryComplete(sequence);
             // update the bool lazily
             // only consider positive limits / negative ones imply tail which means having to go
             // through the whole page of results before selecting the last ones
             // doing a limit early returns the 'head' not 'tail'
-            headLimit = limit != null && limit.limit() > 0 && completed.size() == limit.totalLimit();
+            calculateHeadLimit();
         } else {
             stageToKeys.add(stage, key);
             keyToSequences.add(stage, sequence);
         }
     }
 
+    public void tryComplete(Sequence sequence) {
+        for (int i = 0; i < missingEvents.length; i++) {
+            if (missingEvents[i]) {
+                toCheckForMissing.add(sequence);
+                return;
+            }
+        }
+        completed.add(sequence);
+    }
+
+    private void calculateHeadLimit() {
+        headLimit = limit != null && limit.limit() > 0 && completed.size() == limit.totalLimit();
+    }
+
+    int previousPositiveStage(int stage) {
+        for (int i = stage - 1; i >= 0; i--) {
+            if (missingEvents[i] == false) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    public boolean limitReached() {
+        calculateHeadLimit();
+        return headLimit;
+    }
+
+    public long maxSpanInNanos() {
+        return maxSpanInNanos;
+    }
+
+    int nextPositiveStage(int stage) {
+        for (int i = stage + 1; i < missingEvents.length; i++) {
+            if (missingEvents[i] == false) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    public boolean isMissingEvent(int stage) {
+        if (stage < 0 || stage >= missingEvents.length) {
+            return false;
+        }
+        return missingEvents[stage];
+    }
+
+    private boolean isFirstPositiveStage(int stage) {
+        return stage == firstPositiveStage();
+    }
+
+    int firstPositiveStage() {
```

Comment:
Instead of this method called from various places every time and unless I'm missing something regarding an update to `missingEvents`, firstPositiveStage can be a simple `int`. The same goes for `lastPositiveStage`.